### PR TITLE
Display load config warning only when protected configs in list

### PR DESCRIPTION
**Pull request by mihai-stfc**
### Description of work

Changed the Load Configuration dialogue so that there is no protected configurations warning message when there are none in the list. 
If at least one protected configuration is in the dialogue list, then the message will be displayed.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/5062

### Acceptance criteria

- Warning message and icon are shown only when there is a protected config in the selection list

### Unit tests

N/A

### System tests

N/A

### Documentation
N/A

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 1]
```diff
@@ -148,6 +148,36 @@ public String apply(ConfigInfo info) {
         
         return namesWithProtectionFlag;
     }
+    
+    /**
+     * Returns Configuration/Component names and their protection status, excluding the current configuration
```
Comment (RaiBishal) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:153]: Javadoc requires the first sentence to end with a full stop.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 2]
```diff
@@ -148,6 +148,36 @@ public String apply(ConfigInfo info) {
...
+     * @param infos
+     * @return collection of pair config/comp Name and Protection Flag (e.g. {<name>=<prot. status>, ...}), excluding the current configuration
+     */
+    public static Map<String, Boolean> mapNamesWithTheirProtectionFlagWithoutCurrent(Collection<ConfigInfo> infos) {
```
Comment (RaiBishal) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:157]: Not sure if this method is used anywhere else.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 3]
```diff
@@ -148,6 +148,36 @@ public String apply(ConfigInfo info) {
...
+    	Map<String, Boolean> filteredResults = mapNamesWithTheirProtectionFlag(infos);
+    	filteredResults.remove(Configurations.getInstance().display().displayCurrentConfig().getValue().name());
+    	return filteredResults;
+    }
+    
+    /**
+     * Checks if there is a protected configuration/comp within the given ConfigInfos objects
```
Comment (RaiBishal) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:164]: same as above.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 4]
```diff
@@ -148,6 +148,36 @@ public String apply(ConfigInfo info) {
...
+     * @param infos
+     * @return boolean: True if a protected configuration exists within given infos, False if not
+     */
+    public static Boolean hasProtectedConfigs(Collection<ConfigInfo> infos) {
+        if (infos == null) {
```
Comment (RaiBishal) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:169]: Infos will never be null as it is always assigned in the constructor. However, It can be empty, so we should better check if it is empty.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 5]
```diff
@@ -148,6 +148,36 @@ public String apply(ConfigInfo info) {
...
+            return false;
+        }
+        
+        for (ConfigInfo config: infos) {
+        	if (config.isProtected == true) {
+        		return true;
+        	}
+        }
+    	
+    	return false;
```
Comment (RaiBishal) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:179]: Good to have one return statement rather than 3. Easier to debug If something goes wrong.

**Review by RaiBishal — CHANGES_REQUESTED** [review]
Some minor changes.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 6]
```diff
@@ -148,6 +149,27 @@ public String apply(ConfigInfo info) {
         
         return namesWithProtectionFlag;
     }
+    
+    /**
+     * Checks if there is a protected configuration/comp within the given ConfigInfos objects.
+     * @param infos
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:155]: Please give a description of infos

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 7]
```diff
@@ -148,6 +149,27 @@ public String apply(ConfigInfo info) {
...
+     * @return boolean: True if a protected configuration exists within given infos, False if not
+     */
+    public static Boolean hasProtectedConfigs(Collection<ConfigInfo> infos) {
+        if (infos.isEmpty()) {
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:159]: Suggestion: Use infos.stream().anyMatch(<predicate>) where predicate is the test for isProtected. See here for more details on anymatch https://www.geeksforgeeks.org/stream-anymatch-java-examples/. Stream can be a useful tool for situations like this

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 8]
```diff
@@ -120,10 +120,11 @@ public String synoptic() {
      * @return The list of config names
      */
     public static Collection<String> names(Collection<ConfigInfo> infos) {
-        if (infos == null) {
+        if (infos.isEmpty()) {
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:123]: Though I agree we should check for emptiness and ideally wouldn't need null checks, I do think we should not remove null checks when we can't guarantee how `infos` is made. I suggest `if (infos == null || infos.isEmpty())` as the `||` is cutting so if infos is null then the check `infos.isEmpty()` will not be performed. Please do similar for the other places that we have removed the null check.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java (hunk 7 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:159]: If you choose to implement this suggestion then there is no need for the empty test, but a null check would be useful

File: base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/dialogs/MultipleConfigsSelectionDialog.java [hunk 9]
```diff
@@ -124,20 +124,27 @@ protected void createSelection(Composite container) {
...
-            messageImageLabel.setImage(JFaceResources.
-                getImage(DLG_IMG_MESSAGE_WARNING)); 
-
-        Label messageLabel = new Label(group, SWT.NONE);
-        messageLabel.setFont(JFaceResources.getDialogFont());
-        messageLabel.setText("Represents Protected " + getTypeString());
+		// show protected configuration message only if relevant (the list has protected configurations)
+		Boolean hasProtectedConfigs = ConfigInfo.hasProtectedConfigs(available);
+		if (hasProtectedConfigs)
+		{
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/dialogs/MultipleConfigsSelectionDialog.java:135]: Checkstyle may complain if the `{` is not on the same line as the if with a space between the `{` and the `)`

File: base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/dialogs/MultipleConfigsSelectionDialog.java [hunk 10]
```diff
@@ -124,20 +124,27 @@ protected void createSelection(Composite container) {
...
+			Group group = new Group(container, SWT.SHADOW_IN);
+			group.setLayout(new RowLayout(SWT.HORIZONTAL));
+			        
+			
+		    Label messageImageLabel = new Label(group, SWT.NONE);
+	            messageImageLabel.setImage(JFaceResources.
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/dialogs/MultipleConfigsSelectionDialog.java:141]: Checkstyle may complain about the indentation here. The call to setImage should be indented at the same level as the Label declaration

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java [hunk 11]
```diff
@@ -148,6 +149,27 @@ public String apply(ConfigInfo info) {
         
         return namesWithProtectionFlag;
     }
+    
+    /**
+     * Checks if there is a protected configuration/comp within the given ConfigInfos objects.
+     * @param infos
+     * @return boolean: True if a protected configuration exists within given infos, False if not
+     */
+    public static Boolean hasProtectedConfigs(Collection<ConfigInfo> infos) {
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/configuration/ConfigInfo.java:158]: I would like for there to be unit tests on this method

**Review by JamesKingWork — CHANGES_REQUESTED** [review]
See comments

**Review by JamesKingWork — APPROVED** [review]
Looks great!
